### DISCUSSION TITLE
Missing ability to use keyboard layout per application

### DISCUSSION THREAD

**Discussion by prowwid** [linuxdeepin/developer-center#2958] [Features Request & Ideas | 特性请求 & 头脑风暴]
I've installed latest Deepin 15.5 and can't find an option to use different keyboard layouts for different applications. 

For example:
- use **English** in _Slack_
- use **Russian** in _Chrome_ browser
- use **Ukrainian** in _Telegram Messenger_ 

I've used **CentOS** and **Ubuntu** before **Deepin** and there always we an option to use keyboard layouts per application. 

If there is a way to use keyboard layouts per application please give me some instructions because it's not just mine issue - I see questions about how to enable keyboard layouts per application in forums.

It's a really annoying issue for those who speak several languages.

**Comment by hebrewd** [comment]
I'm having the same issue with deepin.
I did try to set xorg configurations for the keyboard for that sake, like, changing the keyboard layout via capslock, but deepin doesn't seem to quite take that.

**Comment by prowwid** [comment]
@hualet or someone from developers,

could you please respond?

**Comment by aiamuzz** [comment]
I know its not what you may be asking ... Deepin OS is a work in progress !!!

... for the most part they have the OS up and running ... they have a lot of subtlety's covered too (one can get to see them across packages like ... deepin movie, deepin terminal to name a few). 

As this is a fast developing OS ... for things not found or missing ... you may just have to draw their attention to it ... in this(your) case the  application specific keyboard alignment ... they have an internal approval mechanism and if it passes their evaluation ... your request may just be added to their list of features for future updates ...

FYI ... an option to switch between keyboards is available in Deepin already ... without breaking a sweat(an application specific pre-configuring that you are seeking is fine) ... this shortcut just about lives up to your need ... it will easily let you switch between keyboard layouts ... just add the KB Layouts of your needs in the keyboard settings ... and you are good to go ... press and hold 'Alt' and use 'Shift' to toggle between the language of your choice and release the 'Alt' ... its a system wide call and can be called into any application.

**Comment by hebrewd** [comment]
@aiamuzz, Thank you for the reply.
Of course, your point is well understood and is valid.
The keyboard layout switcher does work and does so perfectly fine.
Our point is not to shame deepin for that matter, quite the opposite, I value the work and polish that deepin bring to the linux desktop and I wouldn't bother to reply here if that wasn't the case.

The main point was to verify whether or not we are just missing something, maybe we can't configure a feature that was there and if it isn't, to get the developer take on it.
Even when realize that it won't be added tomorrow for that matter. :)

**Comment by aiamuzz** [comment]
exactly ...
the elaborate post was in response to... @prowwid 's ... call/shout to devs ... for a response ...

:) ... Yes its indeed commendable ... the finesse deepin is working towards achieving ... and what it has achieved so far !!! ... hope to see deepin's attitude towards evolving/developing their OS with inputs from the community continues ... multiple features(options) in the simplest of implementations is what Deepin should never lose focus of ... as long as they continue doing that ... packing more and more features ... wouldn't be a surprise if Deepin leaves other distro's behind ...

All Linux distro's are powerful beasts ... user friendly? ... can't say that about all linux distro's !!!

Its the user friendliness that set Ubuntu as the most popular one ... linux mint crept in and moved ahead ... and now my bets are on Deepin to lead the pack ... the power of linux and user friendliness(that leaves Windows and Mac OS biting dust) under its hood !!!

:)

**Comment by prowwid** [comment]
Hello @aiamuzz,

sorry if I sound rude. 

It definitely wasn't a shout. It's just almost a month from issue report with no response so I thought maybe the issue was missed dy devs. 

My goal was to identify either I'm dumb(and hundreds of other Dppein users) and can't configure such small feature or Deepin is missing ability to use keyboard layout per application.

Such feature is so defaultt for me so I was really surprised that I haven't found it in Deepin.
You see, Ukraine is the biggest outsourcing market and we use at least two languages all the time to communicate with each other and to communicate with a foreign client. And it's really uncomfortable to switch languages, for example, moving from Chrome to IDE and then to Slack.

I only wish for developers to add support of  keyboard layout per application to backlog

**Comment by aiamuzz** [comment]
hahahahaha ... no problem ... 

oh ok ... now i get the use case ... for application specific language keyboard.
but i guess ... in all probability the option is not present in the OS ... may be it never crossed the developers mind ... or may be it crossed their mind and its a future feature planned ... no way to tell !!!

response here is a premium ... guess they are busy honing the OS ... :)

**Comment by hualet** [comment]
Sorry guys, have been very busy dealing with some stuff inside of deepin. The feature is considered and stopped before since we thought it's minority and we have way more things to do.  We'll reconsider it but won't be avaialable very soon I think.

**Comment by hebrewd** [comment]
I don't know if another thread is necessary so I'll post it here, but it seems like the greeter doesn't allow the change of keyboard layout change.
So if the system got locked on a different layout than english, you pretty much can't login, (As passwords are in english)

**Comment by hualet** [comment]
@hebrewd please open another issue and assign it to @kirigayakazushin :)

**Comment by prowwid** [comment]
Hello @hebrewd 

if you are talking about login screen than you should have the ability to switch layout by clicking on keyboard icon at the left of the password field.

At least I switch layout this way as hotkeys don't work.

**Comment by iddqdidkfa** [comment]
@hualet I'm sure you know better about task priorities insed Deepin. But it's really important for many users I know to have separate keyboard layouts for each window/application. And it can be a deal breaker for users used Gnome before. Maybe you'll have time to think once more about priority of this freature. Thanks.

**Comment by SladeThe** [comment]
I agree. I'm trying to switch to Deepin, but I really miss this feature.

**Comment by hualet** [comment]
@jouyouyun @snyh

**Comment by OlehBozhko** [comment]
The is the most important feature that stop me switch to Deepin. I'm using intellij idea for developing. When I switch to another app,  I have to switch keyboard layout to write message or input search request to some website search system. After it I'm switching  back to IDE and notise that keyboard shortcuts not working. After few seconds I understand that keyboard layout is wrong and I switching back to US keyboard layout. This make me nervous.

**Comment by snyh** [comment]
@sshutl   Is "keyboad layout per application" means every application can keep their keyboard-layout status if user have manually changed it? 
How you would change the layout? By right click menu on the application's input box?  
When you would change the layout? Is it the keyboard-layout status should be  keep after logout?

We doesn't do it now, because all of developer in deepin doesn't use this feature, so we  need  more information about the detail of this feature, include how should design the UI interactive and internal logical. 

Could you give more information about this?  Some screenshot about the UI interactive is very wellcome, or some design documents.

**Comment by prowwid** [comment]
Hello @snyh 

Maybe this article will help to understand what we need?
[How To Set Language On A Per-App Basis In Windows 10](https://www.addictivetips.com/windows-tips/how-to-set-language-on-a-per-app-basis-in-windows-10/)

From a user perspective, it's really just being able to change language from default either using hotkeys or language menu icon in one application(e.g. Chrome)  and then when switching to another Application (e.g. WhatsApp) having previously set language for that application(default is primary language).
And then going back and forth between applications also change the language to the one that was used in an application last time.

Hope it would help.

With best regards,
Dmitry

**Comment by OlehBozhko** [comment]
@snyh 
Sorry for long response
Here how it's implemented for example in KDE:
![input_](https://user-images.githubusercontent.com/5792962/48479850-7d89ca80-e7d6-11e8-8497-f5ac068520bf.png)
It has 4 variants of keyboard layout switching policy (1).
In First variant (Global) it works like in Deepin DE (keyboard layout changes for all desktop session)
In Fourth variant (Window)  Keyboard Layout Manager persists individual keyboard layout for every window. I guess that Keyboard Layout Manager subscribes to Window Manager (over D-Bus etc...) and persists hash-table of  pairs (Window ID & Keyboard Layout ID). When it receive event from Window Manager that active window changed it checks if this Window ID exist in hash-table it switch to Keyboard Layout of this entry. Otherwise puts new entry (Active windows ID && Default keyboard layout ID (2)). 
For Second and Third variant  it will be the similar logic but hash table keys has to be DesktopID or Application ID. Also need to notice that hash table has to be cleared when session destroyed.

I hope this explanation is clear. Please let me know if it isn't. 
Thank you!

**Comment by snyh** [comment]
@prowwid @sshutl  Thanks for this explain.  

@shule1987 @hualet @wangjia12 @electricface 这个优先级应该提高的，50%以上的社区用户属于国外用户，很可能会遇到类似需求。

**Comment by snyh** [comment]
@hualet @electricface  Can we implement this by 

1. add a switch policy (_SP_) option in DCC
![image](https://user-images.githubusercontent.com/1437166/48659139-81904380-ea87-11e8-85bc-1f5c8fd0f5b1.png)
2. (if _SP==per_ application) record the layout value and current focus application to the layout map table (_LMT_)
3. (if SP==per application) when focus changed, change keyboard layout according by the _LMT_

**Comment by snyh** [comment]
There a similar option in fcitx's configure panel.  But this is for whether IM actived status not for the keyboard layout switch status.

![image](https://user-images.githubusercontent.com/1437166/48667854-ddf07300-eb1a-11e8-8d17-9a3d4a07a760.png)

**Comment by shule1987** [comment]
![default](https://user-images.githubusercontent.com/5448609/48744653-11f6a000-eca3-11e8-9efa-63de496e7f3b.png)

**Comment by OlehBozhko** [comment]
Any updates?

**Comment by snyh** [comment]
The issue has been scheduled to 15.10

> Any updates?

The issue has been scheduled to 15.10.  We expect it to be released as scheduled.

**Comment by hualet** [comment]
Need to reschedule this one since we have no extra manpower to do this at the moment.

**Comment by OlehBozhko** [comment]
> Need to reschedule this one since we have no extra manpower to do this at the moment.

OK! Thank you for update

**Comment by iddqdidkfa** [comment]
> 
> 
> Need to reschedule this one since we have no extra manpower to do this at the moment.

I think the easiest way to implement it - just to make [kbdd](https://github.com/qnikst/kbdd) library work in Deepin.

**Comment by bublik77** [comment]
Hello,
Any update?

**Comment by saintzyo** [comment]
Hi!
@hualet @justforlxz Any updates on this?
Thanks!

**Comment by alirezaimi** [comment]
Hi
any update ?
Thx.

**Comment by alirezaimi** [comment]
Hi
Is there any hack or shortcut until next update ?! 
This problem is very very painful !

**Comment by prowwid** [comment]
Hi @alirezaimi 

Peronally I haven't found one and switched to another OS.

**Comment by alirezaimi** [comment]
@prowwid 
i'm back to my debian with mate desktop too !

**Comment by artemijan** [comment]
any updates? it's 5.11 version now, but it seems to me that there isn't such a feature.

**Comment by OlehBozhko** [comment]
Any chance to get this feature? This issue is 2.5 years old :)

**Comment by obinoby** [comment]
Hi,
I'm just adding my experience. No hard feelings ;)

In Quebec we generally have a Canadian English keyboards but we speak and write French. So we have to configure our keyboard to Canadian multilingual to get French specific chars. Doing so every symbol used in dev are mapped elsewhere or behind complicated key combinations. So I end-up switching from English layout in terminals and dev tools to multilingual in browser or chat tools.
This feature is a must here too. Coming from Gnome-shell I really miss this feature in Deepin.

I hope it will come some day.

Regards,
Ben

**Comment by goalain** [comment]
This is really a pain for me, working in 3 languages most of the day, sometime even 4,having to select and reselect the layout several time per minute is really the main reason for to opt off the lovely deepin.
This is getting even worst by being unable to set a a single key layout switching, in other environments I use the "Super" alone to switch between layouts and I have an indicator on the task bar.

BTW, the current behavior is even worst than on native ms-win systems.

**Comment by meiyixiang** [comment]
This suggestion is recorded and belongs to a more personalized function.

**Comment by meiyixiang** [comment]
I have seen similar features in the latest versions of Ubuntu, but instead of setting them individually for each application, the user will record the status of your keyboard layout when typing with the added keyboard layout for the corresponding application, and will use the last keyboard layout for typing when it is downloaded and enabled

**Reply by DenisBY** [reply]
It's not an Ubuntu feature rather than DM, i.e. KDE.

**Comment by apsyhea** [comment]
So much nonsense written here. This is not a specific function! This function is in any popular and not so distribution kit! This feature is available on Windows and Mac OS. Find me an operating system in which this does not work out of the box. You won't find. Do you know why? Because it's a Deepin bug. By default, the fixit ​​able to do this, both the 4th version and the 5th. This bank in Liping is already 5 years old, judging by the forum, but for some reason the swindlers do not want to fix it.
